SYCL: Rename oneMKL to oneMath

Rbiessy opened the pull request:
oneMKL Interface was moved to `uxlfoundation` and renamed to [oneMath](https://github.com/uxlfoundation/oneMath). This PR updates the SYCL backend to use the new library.

So far the Intel backend has been using the Intel oneMKL directly. After the renaming Intel oneMKL and oneMath have different namespaces so it is much easier to ensure all SYCL backends use oneMath. As a consequence, `INTEL_CPU` and `INTEL_GPU` targets have been introduced to make it possible to use oneMath compile-time dispatcher. The `INTEL` target can still be used but will fallback to the runtime dispatcher which can introduce a small overhead.
Unlike Intel oneMKL, oneMath is not available as a pre-built package. To ensure it is still as easy to use as before oneMath can be fetched and compiled automatically if not provided with `oneMath_DIR`.

The new version of oneMath had CMake improvements which make it possible to properly integrate oneMath with CMake rather than relying on environment variables such as `LIBRARY_PATH` or `CPLUS_INCLUDE_DIR`.

Using `get_onemath_backend` lets us avoid duplicating paths when using oneMath with different backends. The changes will help us ensure we avoid compilation issues when using APIs only available in Intel oneMKL.

I believe the changes will also fix the issues mentioned in https://github.com/ggml-org/llama.cpp/pull/10851 in a cleaner way.

Rbiessy committed:
Rename oneMKL Interface to oneMath

Rbiessy committed:
Use oneMath for Intel vendor

Rbiessy committed:
Rename occurences to mkl

Rbiessy committed:
clang-format

Rbiessy committed:
Silence verbose warnings

Rbiessy commented:
This change should not affect performance for Intel backends as long as the compile-time dispatcher is used. Some performance results before my changes:
| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ---: | ------------: | -------------------: |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        512.45 ± 1.26 |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |          4.90 ± 0.00 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |      3411.37 ± 14.70 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         34.82 ± 0.18 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |      3230.48 ± 21.10 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         21.61 ± 0.10 |

with these changes:
| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ---: | ------------: | -------------------: |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        514.92 ± 3.08 |
| llama 70B Q4_K - Small         |  37.57 GiB |    70.55 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |          4.89 ± 0.00 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |      3399.59 ± 13.79 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         34.74 ± 0.16 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |      3259.86 ± 10.15 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         21.53 ± 0.08 |

s-Nick commented:
Changes look good to me, but when I build it from scratch there are a lot of warnings from oneMath. 
Is there a way to hide them? I think it would provide a better user experience and also it would helps developers to monitor warnings only from llama.cpp

s-Nick commented on docs/backend/SYCL.md:
```diff
@@ -228,27 +228,27 @@ Upon a successful installation, SYCL is enabled for the available intel devices,
+git clone https://github.com/uxlfoundation/oneMath
+cd oneMath
+# Find your HIP_TARGETS with rocminfo, under the key 'Name:'
+cmake -B buildWithrocBLAS -DCMAKE_CXX_COMPILER=icpx -DCMAKE_C_COMPILER=icx -DENABLE_MKLGPU_BACKEND=OFF -DENABLE_MKLCPU_BACKEND=OFF -DENABLE_ROCBLAS_BACKEND=ON -DHIP_TARGETS=${HIP_TARGETS} -DTARGET_DOMAINS=blas  -DCMAKE_INSTALL_PREFIX:PATH=install
```
Since you added `FetchContent_Declare` and the cmake automatically configures oneMath, what do you think about a line somewhere specifying that building oneMath before llama.cpp is not mandatory anymore?

Rbiessy replied:
Indeed, I added a note and also changed the instructions to rely on oneMath being pulled automatically since this is easier for most users: https://github.com/ggml-org/llama.cpp/pull/12192/commits/9002b4decf25973ecb4c28206471f362cddbcca0

Rbiessy commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,93 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
+    function(onemath_alias target)
+        if (TARGET ${target})
+            # Silence verbose warnings from external libraries
+            target_compile_options(${target} PRIVATE -Wno-uninitialized -Wno-unused-parameter -Wno-unused-variable -Wno-cast-qual)
```
```quote
when I build it from scratch there are a lot of warnings from oneMath.
Is there a way to hide them? I think it would provide a better user experience and also it would helps developers to monitor warnings only from llama.cpp
```
@s-Nick I thought this line fixed the warnings you mentioned but I probably missed that I need to compile it from scratch. I'll have a look.

abhilash1910 replied:
Maybe adding a `-w` at the start can help ? LGTM otherwise.

Rbiessy replied:
This is fixed in https://github.com/ggml-org/llama.cpp/pull/12192/commits/bc851c856867a9f80ed2f7393e096eb1aa459c92. The flags need to be set on another target. I ended up using `-w` as well.

Rbiessy committed:
Set oneMath HIP_TARGETS

Rbiessy committed:
Fix silence warnings

Rbiessy committed:
Remove step to build oneMath from build instructions

NeoZhangJianyu commented on docs/backend/SYCL.md:
```diff
@@ -291,6 +271,8 @@ For AMD GPUs we should expect at least one SYCL-HIP device [`hip:gpu`]:
 
 ### II. Build llama.cpp
 
+The SYCL backend depends on [oneMath](https://github.com/uxlfoundation/oneMath). By default it is automatically built along with the project. A specific build can be provided by setting the CMake flag `-DoneMath_DIR=/path/to/oneMath/install/lib/cmake/oneMath`.
```
For Intel GPU, we still use Intel oneMKL included in oneAPI package now.
For non Intel GPU, maybe use pure open source oneMath.
I suggest not to impact Intel GPU in this PR, including code and document.

Rbiessy replied:
One of the goal of this PR is also to unify the libraries used for the different SYCL backends. oneMath was introduced for this purpose.
We have seen issues before where a namespace used compiled when using Intel oneMKL but not oneMKL Interface (i.e. oneMath). I have introduced the `FetchContent` to avoid any impact on the user and from my benchmarks this does not affect performance either.

NeoZhangJianyu replied:
SYCL backend use oneDNN to replace oneMKL to implement mul_mat() functions as default.
it's always ignored.

If don't change the related code, the oneMath/oneMKL code path won't be tested.

NeoZhangJianyu reviewed (changes requested):
oneMath is new & good product for multiple platform.
It's good to be used to support multiple GPUs.

Because SYCL backend has been used in some commercial products for Intel GPU, we must provide stable solution based on stable comments.

So, I think for Intel GPU, we still continue using components provided by official oneAPI: compiler, oneMKL, oneDNN.

If you want to use oneMath right now, please use Macro to define new branch for non Intel GPU code path.
When it's done, please build and verify on pure oneAPI environment for Intel GPU.

Additionally, please consider if need to support Intel CPU in SYCL backend.
I don't think it bring benefit to Intel CPU user than CPU backend.

Thank you!

NeoZhangJianyu commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -1686,26 +1709,18 @@ namespace dpct
-                                                  beta_value, data_c, ldc);
-#endif
-        }
+    template <class Ta, class Tb, class Tc, class Ts>
```
I suggest warp gemm() functions of oneMKL and oneMath in one unify API() in dpct::helper, so that make the ggml code is clear & simple.

NeoZhangJianyu replied:
For Intel path, I suggest using oneMKL which comes from oneAPI official package, instead of oneMath as 3rd party.

If define a unified API, like dpct::gemm(dcpt::transpose a_trans, xxxx, int ldc), this new function will hide the difference of oneMath and oneMKL. 

That means ggml-sycl.cpp won't call oneMKL or oneMKL directly, only dpct::helper call oneMKL or oneMath.
So when we change low level code, ggml-sycl.cpp won't be impacted directly.

qnixsynapse reviewed (approved):
Tested on an Intel A750 with intel-oneapi-base-toolkit 2025.0.1, works okay..

LGTM.


Edit: Also, we should remove the mention of Intel CPU from the docs as this backend only targets GPUs.

NeoZhangJianyu commented:
I tested the code in pure oneAPI. It's passed. It's good!

I don't oppose to add new packages for non-intel GPU. I understand you.
I hope SYCL backend for Intel GPU use stable official packages.

I suggest divide the code of intel GPU and non Intel GPU in SYCL backend if you want to add oneMath and more packages:

In code level, I have two alternatives:
1. Design a new class to provide unify API for gemm() functions based on oneMKL, or oneMatch or oneDNN. (recommend)
   for intel, call oneMKL or oneDNN like base.
   for non-intel, call oneMath.
   
   In CMakefile.txt, I suggest design two branches for intel and non-intel. 
   So you only update the non-intel branch.

2. Divide SYCL backend into two: SYCL backend, Multiple SYCL backend.
  - SYCL backend support Intel device only. 
  - Multiple SYCL backend support Intel/NV/AMD devices.
  These two backends are maintained separately and less conflict.
  They could refer to each other's code and design.
  The code could be clean and compact. In current code, there are branches for intel or non-intel. And they will be more with the more optimization in the future.


Thank you!

NeoZhangJianyu commented:
Another concern is the CMake will download the oneMath during building.
That will impact the build speed.

Is it possible to move the download from CMake to prepare work like install oneAPI?

Thank you!

airMeng commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,99 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
+
+find_package(oneMath QUIET)
+if (NOT oneMath_FOUND)
+    message("-- oneMath not found: oneMath will be automatically downloaded")
```
better wait for oneMath be part of oneAPI and more stable then do the porting for intel GPU

Rbiessy replied:
oneMath will not be part of the oneAPI base toolkit. It is part of the UXL foundation.
The blas API in particular has already been very stable for more than a year. The renaming to oneMath was designed to avoid breaking changes as well. Do you have a concrete example to show the project is not stable enough?

Rbiessy commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,99 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
+    endif()
+    FetchContent_Declare(
+        ONEMATH
+        GIT_REPOSITORY https://github.com/uxlfoundation/oneMath.git
```
Will do

Rbiessy replied:
Done in https://github.com/ggml-org/llama.cpp/pull/12192/commits/09dfe8981d7d75ec7e0f53b652f2138afbd039f3

Rbiessy commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,99 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
-        add_compile_definitions(GGML_SYCL_NVIDIA)
-        target_link_libraries(ggml-sycl PRIVATE sycl pthread m dl onemkl_blas_cublas)
+    # Fallback to the simplest way of enabling SYCL when using intel/llvm nightly for instance
+    target_compile_options(ggml-sycl PRIVATE "-fsycl")
```
Yes these flags should not be needed. CMake takes care of forwarding all the flags needed when linking against a target such as `ONEMATH::onemath_blas_mklgpu` with `target_link_libraries`. Applications should not manually do this.

I have run this PR on Intel GPU, the results in this comment where run on PVC: https://github.com/ggml-org/llama.cpp/pull/12192#issuecomment-2700565080

Rbiessy commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,99 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
+    target_link_libraries(ggml-sycl PRIVATE ONEMATH::onemath_blas_mklcpu)
+    target_compile_definitions(ggml-sycl PRIVATE GGML_SYCL_INTEL GGML_SYCL_INTEL_CPU)
+elseif (GGML_SYCL_TARGET STREQUAL "INTEL_GPU")
+    target_link_libraries(ggml-sycl PRIVATE ONEMATH::onemath_blas_mklgpu)
```
I don't see a reason to only use the Intel oneMKL package provided by the oneAPI base toolkit. The SYCL backend is meant to be a generic backend that can support multiple HW so we should avoid splitting for different paths as long as we can maintain performance. oneMath is designed to solve this issue exactly.

Rbiessy commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -83,13 +83,36 @@ inline std::string get_device_backend_and_type(const sycl::device &device) {
 
 template <typename Ts> struct matrix_info_t {
-    oneapi::mkl::transpose transpose_info[2];
+    oneapi::math::transpose transpose_info[2];
```
Resolving the comment since this is a duplicate of your other comment: https://github.com/ggml-org/llama.cpp/pull/12192#discussion_r1984359269

Rbiessy commented:
@NeoZhangJianyu @airMeng there are a few comments regarding concerns of introducing oneMath for Intel devices.

Pasting my same reply from one of the comment: The SYCL backend is meant to be a generic backend that can support multiple HW so we should avoid splitting for different paths as long as we can maintain performance. oneMath is designed to solve this issue exactly.

```quote
Another concern is the CMake will download the oneMath during building.
```
Using FetchContent will download oneMath during CMake configuration time already, see https://cmake.org/cmake/help/latest/module/FetchContent.html. A user can also choose not to use FetchContent by building oneMath separately and installing it or setting `-DoneMath_DIR`.

Rbiessy committed:
Use fixed oneMath version

Rbiessy committed:
Remove INTEL_CPU

Rbiessy committed:
Fold CMake oneDNN conditions

Rbiessy commented on docs/backend/SYCL.md:
```diff
@@ -659,7 +630,7 @@ use 1 SYCL GPUs: [0] with Max compute units:512
 |--------------------|---------------------------------------|---------------------------------------------|
 | GGML_SYCL          | ON (mandatory)                        | Enable build with SYCL code path.<br>FP32 path - recommended for better perforemance than FP16 on quantized model|
-| GGML_SYCL_TARGET   | INTEL *(default)* \| NVIDIA \| AMD    | Set the SYCL target device type.            |
+| GGML_SYCL_TARGET   | INTEL *(default)* \| INTEL_CPU \| INTEL_GPU \| NVIDIA \| AMD    | Set the SYCL target device type.            |
```
Removed mentions of INTEL_CPU in https://github.com/ggml-org/llama.cpp/pull/12192/commits/948f3c5f1e37bd9d548efed15dc775a358d7fc58

Rbiessy commented on examples/sycl/README.md:
```diff
@@ -14,7 +14,7 @@ List all SYCL devices with ID, compute capability, max work group size, ect.
 1. Build the llama.cpp for SYCL for the specified target *(using GGML_SYCL_TARGET)*.
 
-2. Enable oneAPI running environment *(if GGML_SYCL_TARGET is set to INTEL -default-)*
+2. Enable oneAPI running environment *(if GGML_SYCL_TARGET is set to INTEL -default-, INTEL_CPU or INTEL_GPU)*
```
Removed in https://github.com/ggml-org/llama.cpp/pull/12192/commits/948f3c5f1e37bd9d548efed15dc775a358d7fc58

Rbiessy commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,99 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
 endif()
 
-if (${DNNL_FOUND} AND GGML_SYCL_TARGET STREQUAL "INTEL")
+if (${DNNL_FOUND} AND GGML_SYCL_TARGET MATCHES "INTEL.*")
```
Removed in https://github.com/ggml-org/llama.cpp/pull/12192/commits/948f3c5f1e37bd9d548efed15dc775a358d7fc58

Rbiessy commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -51,36 +49,99 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl-targets=nvptx64-nvidia-cuda")
-        add_compile_definitions(GGML_SYCL_NVIDIA)
-        target_link_libraries(ggml-sycl PRIVATE sycl pthread m dl onemkl_blas_cublas)
+    # Fallback to the simplest way of enabling SYCL when using intel/llvm nightly for instance
```
The existing CMake already allowed to use intel/llvm nightly. I have just clarified this with a comment.
It can be useful for development purposes or for users that build intel/llvm themselves.

Rbiessy commented:
@NeoZhangJianyu @airMeng I believe I have addressed all the comments. Do let me know if you have more concerns with the approach that I haven't answered. If not please do not keep the PR blocked, thanks!

NeoZhangJianyu commented:
Thank your patience to answer the comments! :)
1. I think oneMath is immature product now.
2. I don't oppose to use it in non-Intel GPU case.
  If the oneMath is only used in non-Intel GPU case, I agree to merge it.

Here is the reason:
For 1:
  oneMath's latest release is 0.6 on No.7 Nov. 2024.
  There are still warning in building oneMath log in CI.
  oneMath can't cover all APIs of oneMKL by now.  That means we can't use some existed API of oneMKL by oneMath API.

For 2:
  1. The only benefit I learn is reduce the code branch of oneMKL in Intel and non-Intel case.

  2. Side-effect:

    a. need more building time:
           It need to build from source code when using it in llama.cpp.
          Including download and build it, the building time is increased from 21s to 42s in linux.

    b. create more lib files:
         In windows, it create 3 dll files in build/bin:
             557056      onemath.dll
            121745920 onemath_blas_mklcpu.dll
            121745920 onemath_blas_mklgpu.dll

             It take more time when link to new dll files, but I didnt record.

         In liunx, it create 3 files in `build/lib`:
             2703600  libonemath_blas_mklcpu.so.0
             2703600  libonemath_blas_mklgpu.so.0
             914544    libonemath.so.0


      I don't know if they need to be included in llama.cpp release package.
      If yes, it means more disk cost.

    c. break the offline build style of SYCL backend.
       Through oneMath is downloaded during make automatically, it still add dependence.
       Some developer access github unstable by internet.
       Additional download will increase the risk of CI and build time.

    d. Add new warnings of oneMath.

I think oneMath is good product to reduce the multiple APIs cross platforms.
But it's not good enough now.
SYCL backend is used by some commercial developers for Intel GPU, instead of non-Intel GPU.
That's why I always high light to use the stable official packages (like oneAPI toolkit).

For non-Intel GPU, oneMath is OK. Maybe nobody care above side effect.

Above is all I want to say.

If you still like to merge this PR, I could approve.
But could I will create another PR to remove the side effort above of Intel GPU by skip using oneMath for Intel GPU?

By the way, when we create SYCL backend, Intel GPU is only target device. It's same of mine now.
I contribute to llama.cpp as personal contributor in my spare time.
It's due to my interesting and I hope more Intel GPU users can enjoy llama.cpp.


Thank you!

NeoZhangJianyu commented:
@zunigasllc why spam join this review?
Is it so hot PR? :)

qnixsynapse commented:
@NeoZhangJianyu please ignore, it's a spam account.

Rbiessy committed:
Merge branch 'master' into romain/rename_onemath

Rbiessy commented:
@NeoZhangJianyu, thank you for the feedback.

Regarding your point 1 the CI only see a few warnings in the tests on linux related to the random number generation. They are harmless and not enough to consider the library not stable. oneMath is only used for GEMM and batched GEMM. As far as I know all the features provided by Intel oneMKL for these operations are also available in oneMath. If that's not the case we can help integrate the missing features in oneMath.

Regarding the side effects in your point 2, the main benefit is indeed to reduce the code branching. I believe this helps SYCL developers in general. It would also help avoid issued such as [this one](https://github.com/ggml-org/llama.cpp/pull/10257/files#diff-1d8f6534cb36440274249e08d33ec23eb8431224d0d5526907fd58d49d5258f5R43) where a namespace used is only available in Intel oneMKL but not oneMath.
**a. & b.** I agree there is a cost for the building time and disk space (just 2.6MB from my measurements). Note that with the recent changes the libraries `libonemath_blas_mklcpu.so` and `libonemath.so` should not be built anymore. With ccache enabled the compilation time will be barely measurable.
**c.** As documented in https://github.com/ggml-org/llama.cpp/pull/12192/files#diff-4f40b10d65d19565440cfa25acd0129a87a596de19c2298991021a9a01a040e6R274 users can still download, build and install oneMath separately if it is important for them. Yes it is one more dependence, that is the solution that has been designed by Intel to provide generic SYCL packages. Note that we have also tried to push for oneMath to be available as a pre-built packages in the discussions [here](https://github.com/uxlfoundation/oneMath/issues/467). This is not a priority at the moment but integrating oneMath in llama would be one more argument to enable this. It is a chicken-egg problem, we can't improve oneMath if it is not used anywhere and we can't use it in projects if it is not "good enough". From our experience with oneMath it is indeed mature enough, it is being used in [GROMACS](https://gitlab.com/gromacs/gromacs/-/blob/main/docs/install-guide/index.rst?plain=1#L455).
**d.** I have fixed the warnings showing up. Llama enable strict warnings globally, these have been disabled for oneMath. As mentioned above, the Intel CI does not show warnings with the default warnings for the library itself. See for instance this [job](https://github.com/ggml-org/llama.cpp/actions/runs/13842084824/job/38731793313?pr=12192) from this GitHub CI running.


```quote
If you still like to merge this PR, I could approve.
```
Yes it would really help us if you don't block such PRs, thanks. Your feedback is welcomed though!

```quote
But could I will create another PR to remove the side effort above of Intel GPU by skip using oneMath for Intel GPU?
```
If customers face issues I think we should first look into a solution to fix them in oneMath. If it is not possible we could consider reverting to using Intel oneMKL directly. Until then I would like to try using oneMath.

I hope this answers all of the remaining concerns.

airMeng commented:
how about the performance on BMG/LNL? The performance number you show is only on PVC.

NeoZhangJianyu commented:
@Rbiessy 
I still think Intel GPU won't be impacted by the PR (oneMath).
In this year, the optimization gate will opened, but Intel, CUDA/Hip can't walk on same path:
It's impossible to make SYCL backend get good performance on Intel, NV & AMD GPU by same code.

SYCL backend is more important to Intel user and ISV on Intel GPU.
We must make sure this path is always stable and fast.

I hope allow me create another PR to skip oneMath for Intel GPU.

Rbiessy commented:
> how about the performance on BMG/LNL? The performance number you show is only on PVC.

@airMeng I did not run them before as I was confident enough it would not show up there either. The results are below for a couple of small models. The differences are just noise.

<details>
<summary>master results on BMG</summary>

| model                          |       size |     params | backend    | ngl |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ----: | ---: | ------------: | -------------------: |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         pp512 |       2074.35 ± 2.41 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         tg128 |         31.75 ± 0.16 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         pp512 |     1921.31 ± 335.13 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         tg128 |         10.55 ± 0.02 |
</details>


<details>
<summary>PR results on BMG</summary>

| model                          |       size |     params | backend    | ngl |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ----: | ---: | ------------: | -------------------: |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         pp512 |       2075.38 ± 4.72 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         tg128 |         32.80 ± 0.18 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         pp512 |     1897.90 ± 277.29 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |  none |    0 |         tg128 |         10.63 ± 0.01 |
</details>


<details>
<summary>master results on LNL</summary>

| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ---: | ------------: | -------------------: |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        534.03 ± 1.25 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         12.10 ± 0.85 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        548.45 ± 4.34 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |          3.86 ± 0.01 |
</details>

<details>
<summary>PR results on LNL</summary>

| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ---: | ------------: | -------------------: |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        535.93 ± 2.77 |
| llama 8B Q4_K - Medium         |   4.58 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |         11.84 ± 1.00 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         pp512 |        574.42 ± 2.25 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | SYCL       |  99 |       8 |  none |    0 |         tg128 |          3.99 ± 0.19 |
</details>

Rbiessy commented:
>  Intel, CUDA/Hip can't walk on same path: It's impossible to make SYCL backend get good performance on Intel, NV & AMD GPU by same code.
> 
> SYCL backend is more important to Intel user and ISV on Intel GPU. We must make sure this path is always stable and fast.

@NeoZhangJianyu We agree on this. We are not trying to make all backends use the same paths for every operation. For the common fp16 or fp32 gemm it is possible because all vendors provide native libraries with similar APIs. We also want to make sure the performance is optimal on Intel devices. This is just helping us clean some code.
From what we have seen so far oneMath operations are not that much used, in particular if oneDNN is enabled. Hence I think you are making this a bigger problem than what is really. The biggest challenge to improve the SYCL backend performance today is to optimize the quantized matmul variants which do not rely on oneMath nor oneDNN.

> I hope allow me create another PR to skip oneMath for Intel GPU.

As I said above, if there are important issues first let us know what they are and we can discuss the solutions then. We could find a solution to allow using Intel oneMKL again if needed.

NeoZhangJianyu commented:
I have listed the side-effect of this PR.

airMeng reviewed (approved):
Looks good on BMG/LNL, waiting for more feedback from the user side

Rbiessy commented:
For context @NeoZhangJianyu found that oneMath is a large library on Windows of around 100MB. This is due to static linking, I created an issue about it here: https://github.com/uxlfoundation/oneMath/issues/654

I agree this is a valid concern so I'll revert to using Intel oneMKL directly for Intel devices.

Rbiessy committed:
Use Intel oneMKL for Intel devices

Rbiessy committed:
Merge branch 'master' into romain/rename_onemath

Rbiessy committed:
Improve CMake message

Rbiessy commented:
@NeoZhangJianyu I've switched back to using Intel oneMKL for Intel devices in https://github.com/ggml-org/llama.cpp/pull/12192/commits/995aea3d8309f049894aaffbc98cd8e774477b3e

Let me know if you have any concern with this approach, thanks.

NeoZhangJianyu commented on docs/backend/SYCL.md:
```diff
@@ -300,6 +280,8 @@ For AMD GPUs we should expect at least one SYCL-HIP device [`hip:gpu`]:
 
 ### II. Build llama.cpp
 
+The SYCL backend depends on [oneMath](https://github.com/uxlfoundation/oneMath). By default it is automatically built along with the project. A specific build can be provided by setting the CMake flag `-DoneMath_DIR=/path/to/oneMath/install/lib/cmake/oneMath`.
```
This description is not mandatory for Intel GPU.
Suggest move to NV/AMD GPU chatpers.

Rbiessy replied:
Moved in https://github.com/ggml-org/llama.cpp/pull/12192/commits/06fe2ca7ed720521664edc504d2a0723a6a92039

NeoZhangJianyu commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -76,34 +91,84 @@ else()
+        onemath_alias(onemath_blas_cublas)
+        onemath_alias(onemath_blas_rocblas)
     endif()
-    if (GGML_SYCL_TARGET STREQUAL "INTEL")
-        target_link_libraries(ggml-sycl PRIVATE sycl OpenCL mkl_core pthread m dl mkl_sycl_blas mkl_intel_ilp64 mkl_tbb_thread)
```
This two lines should be kept:
They will make the SYCL backend depends on the minimum MKL lib files.
Otherwise, SYCL binary will need more MKL so files which are not used in fact.

NeoZhangJianyu commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -76,34 +91,84 @@ else()
+        target_compile_definitions(ggml-sycl PRIVATE GGML_SYCL_AMD)
+    else()
+        # Fallback to oneMath runtime dispatcher
+        target_link_libraries(ggml-sycl PRIVATE ONEMATH::onemath)
```
This path is for Intel in fact.
Please update to use oneMKL .

Rbiessy replied:
Intel devices will not go through this path. This whole section is in the `else()` branch from https://github.com/ggml-org/llama.cpp/pull/12192/files/0b6f9a90978e79684de0fb5d9129e8f7e2bddcfb#diff-123d14cf628b04694c9022d4210d80c14a29409095635150b944f8dfd1c70b37R105

NeoZhangJianyu commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -83,13 +92,32 @@ inline std::string get_device_backend_and_type(const sycl::device &device) {
 
 template <typename Ts> struct matrix_info_t {
-    oneapi::mkl::transpose transpose_info[2];
+    oneapi::math::transpose transpose_info[2];
```
Should it rollback to "mkl" for Intel GPU?

Rbiessy replied:
I have added a namespace alias in https://github.com/ggml-org/llama.cpp/pull/12192/files/0b6f9a90978e79684de0fb5d9129e8f7e2bddcfb#diff-184c9a4290e9a1e26836c54b09c1de9bbf609c50564f21b19fac18f7b74ba25fR25. This means that anything available in `oneapi::mkl` will also be available in `oneapi::math`. This makes it much easier to use both oneMath and Intel oneMKL without more branching everywhere.
`oneapi::mkl` is a deprecated namespace in oneMath so we should avoid using this namespace.

NeoZhangJianyu commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -1686,26 +1714,18 @@ namespace dpct
+        auto data_a      = get_memory<const Ta>(a);
+        auto data_b      = get_memory<const Tb>(b);
+        auto data_c      = get_memory<Tc>(c);
+        oneapi::math::blas::column_major::gemm(get_onemath_backend(q), a_trans, b_trans, m, n, k, alpha_value, data_a,
```
Should it rollback to "mkl" for Intel GPU?

NeoZhangJianyu commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -1735,7 +1755,7 @@ namespace dpct
 
         template <class Ta, class Tb, class Tc, class Ts>
-        inline void gemm_batch_impl(sycl::queue & q, oneapi::mkl::transpose a_trans, oneapi::mkl::transpose b_trans,
+        inline void gemm_batch_impl(sycl::queue & q, oneapi::math::transpose a_trans, oneapi::math::transpose b_trans,
```
Should it rollback to "mkl" for Intel GPU?

NeoZhangJianyu commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -1754,48 +1774,28 @@ namespace dpct
-                matrix_info->ld_info + 1, reinterpret_cast<Ts *>(matrix_info->value_info + 1),
-                reinterpret_cast<Tc **>(c), matrix_info->ld_info + 2, 1, &(matrix_info->groupsize_info));
-#endif
+            sycl::event e = oneapi::math::blas::column_major::gemm_batch(
```
Should it rollback to "mkl" for Intel GPU?

NeoZhangJianyu commented on ggml/src/ggml-sycl/dpct/helper.hpp:
```diff
@@ -1754,48 +1774,28 @@ namespace dpct
-                                                        stride_a, data_b, ldb, stride_b, beta_value, data_c, ldc,
-                                                        stride_c, batch_size);
-#endif
+            oneapi::math::blas::column_major::gemm_batch(get_onemath_backend(q), a_trans, b_trans, m, n, k, alpha_value,
```
Should it rollback to "mkl" for Intel GPU?

NeoZhangJianyu reviewed (commented):
I have added several small comments.
There is a common issue in all changed CPP files:    
   There are still calling oneapi::math for Intel GPU.
   Should it rollback to "mkl" for Intel GPU?

Rbiessy committed:
Link against MKL::MKL_SYCL::BLAS only

Rbiessy committed:
Move oneMath documentation to Nvidia and AMD sections

NeoZhangJianyu reviewed (approved):
It's great work!
Thank you!
